File: Tree/307. Range Sum Query - Mutable.cpp



class NumArray {
  int n;
  vector<int> tree;
  public:
    void buildTree(vector<int>& nums) {
      for (int i = 0; i < nums.size(); i++) {
        tree[n + i] = nums[i];
      }

      for (int i = n - 1; i >= 0; i--) {
        tree[i] = tree[2 * i] + tree[2 * i + 1];
      }
      // done building already
    }

    int query(int left, int right) {
      int res = 0;
      left += n;
      right += n;
      while (left <= right) {
        if (left % 2 == 1) {
          res += tree[left];
          left++;
        }
        if (right % 2 == 0) {
          res += tree[right];
          right--;
        }

        left/=2;
        right /=2;
      }
      return res;

    }





    NumArray(vector<int>& nums) {
      n = nums.size();
      tree.resize(2 * n, 0);
      buildTree(nums);

    }

    void update(int index, int val) {

      index += n;
      int left = index;
      int right = index;
      tree[index] = val;
      while (index > 0) {
        left = right = index;
        if (index % 2 == 0) {
          right++;
        } else {
          left--;
        }
        tree[index/2] = tree[left] + tree[right];
        index /= 2;


      }
      return;
    }

    int sumRange(int left, int right) {
      return query(left, right);


    }
};

File: Tree/307. Range Sum Query - Mutable.py
class Tree:
    def __init__(self, length, array):
        self.tree = [0 for i in range(length)]
        self.n = length // 2
        for i in range(len(array)):
            self.tree[len(array) + i] = array[i]
        for i in range(length - 1 - len(array), 0, -1):
            self.tree[i] = self.tree[2 * i] + self.tree[2 * i + 1]

    def query(self, left: int, right: int) -> int:
        res = 0
        left += self.n
        right += self.n
        while left <= right:
            if left % 2 == 1:
                res += self.tree[left]
                left += 1
            if right % 2 == 0:
                res += self.tree[right]
                right -= 1
            left //= 2
            right //= 2

        return res

    def update(self, index, val):
        index += self.n
        self.tree[index] = val

        # Propagate the change up the tree
        while index > 0:
            left = right = index
            if index % 2 == 0:
                right = index + 1
            else:
                left = index - 1
            self.tree[index // 2] = self.tree[left] + self.tree[right]
            index //= 2


class NumArray:
    def __init__(self, nums: List[int]):
        n = len(nums)
        self.tree = Tree(2 * n, nums)

        return

    def update(self, index: int, val: int) -> None:
        self.tree.update(index, val)

    def sumRange(self, left: int, right: int) -> int:
        return self.tree.query(left, right)

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)